### PR TITLE
WB-623: Implement single select OptionItems in Dropdown 

### PR DESCRIPTION
**Pull request by kumailn**
Added the ability to override the opened state of Dropdown and implemented a new prop to treat OptionItems as single selects (select only one & close dropdown when clicked)

### CODE REVIEW

**Comment by jeresig** [comment]
Deploy preview for *wonder-blocks* ready!

Built with commit 17cf8ae872310ca2fd9e05e3ce7ea8a77ec59829

https://deploy-preview-448--wonder-blocks.netlify.com

File: packages/wonder-blocks-dropdown/components/dropdown.js [hunk 1]
```diff
@@ -34,6 +34,12 @@ type Props = {|
      */
     menuItems: Array<Item> | Item,
 
+    /**
+     * Closes the Dropdown when an OptionItem is selected, use this
+     * prop if you want single-select OptionItems
+     */
+    singleSelectOption?: boolean,
```
Comment (jandrade) [inline packages/wonder-blocks-dropdown/components/dropdown.js]: What do you think about using `singleSelectEnabled` or `singleSelectionEnabled` for consistency with booleans.

To give you some context, there are a lot of props that don't use this naming, but we are trying to get a more consistent API naming, so it would be helpful to name new props this way :)

File: packages/wonder-blocks-dropdown/components/dropdown.js [hunk 2]
```diff
@@ -101,6 +107,12 @@ export default class Dropdown extends React.Component<Props, State> {
         };
     }
 
+    componentDidMount() {
+        if (this.props.opened === true || this.props.opened === false) {
```
Comment (jandrade) [inline packages/wonder-blocks-dropdown/components/dropdown.js]: Is there any particular reason to evaluate opened here? Based on my initial observation, it seems you could remove this condition, but I want to double check with you first.

**Comment by jandrade** [comment]
It's great that you added an example showcasing this new prop. Could you also add a unit test checking this new prop is behaving properly?

File: packages/wonder-blocks-clickable/docs.md [hunk 3]
```diff
@@ -1,4 +1,4 @@
-The Clickable component allows you to easily turn any custom compoent into a clickable one with consistent behavior and accesibility features.
+The Clickable component allows you to easily turn any custom component into a clickable one with consistent behavior and accesibility features.
```
Comment (somewhatabstract) [inline packages/wonder-blocks-clickable/docs.md]: Nice find! I wonder how hard it would be to get some sort of spellchecker thing to run on our docs. Not something for you to tackle, just musing aloud.

File: packages/wonder-blocks-dropdown/components/dropdown.js (hunk 1 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-dropdown/components/dropdown.js]: Could this be even more descriptive of its purpose? Like `closeOnSingleSelect` or `closeOnSingleSelectEnabled`?

While the docs here explain what it does, those docs aren't present at places where this actually gets set and having a more descriptive name will then mean that uses are immediately easier to grok.

File: packages/wonder-blocks-dropdown/components/dropdown.js (hunk 2 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-dropdown/components/dropdown.js]: Could this just be `this.props.opened != null`?

**Comment by somewhatabstract** [comment]
This needs unit tests to validate the updated behavior.

File: packages/wonder-blocks-dropdown/components/dropdown.js (hunk 2 above)
Comment (kumailn) [reply packages/wonder-blocks-dropdown/components/dropdown.js]: @jandrade I wanted to override the opened state only if opened was passed in, I think @somewhatabstract’s suggestion works the same

File: packages/wonder-blocks-dropdown/components/dropdown.js (hunk 1 above)
Comment (kumailn) [reply packages/wonder-blocks-dropdown/components/dropdown.js]: @jandrade I included `Option` in the name because I wanted to convey that this prop only affected the behavior of OptionItems. @somewhatabstract It doesn’t only close the Dropdown however, it also defines the single select behavior by updating the state of the parent component accordingly

Comment (jandrade) [reply packages/wonder-blocks-dropdown/components/dropdown.js]: If that's the case, I'd suggest renaming this prop to something more generic and familiar for devs that previously have worked with `<SingleSelect/>` and `<MultiSelect/>` components. I understand your point about restricting the behavior with the prop name, but I think that maybe it could be better to define that behavior in the value. This way, there would be a mapping between the Dropdown component and the other variants (SingleSelect, MultiSelect and ActionMenu if needed).

Instead of using a boolean here, you could use a union type with values that could help to describe the behavior you are trying to apply.

Here are some examples to try to illustrate my point a bit better:
e.g. 
a. `selectAs: "singleSelect" | "multiSelect"`
b. `type: "single" | "multi"`
c. `selectionType: "single" | "multi"`

..and making one of those options as default.

This way, this prop could be used later without needing to add a new prop for a given behavior.

Any thoughts?

Comment (kevinbarabash) [reply packages/wonder-blocks-dropdown/components/dropdown.js]: I like an enum for this kind of thing.

File: packages/wonder-blocks-dropdown/components/dropdown.md [hunk 4]
```diff
@@ -176,6 +176,7 @@ class MixedDropdownExample extends React.Component {
 
         return (
             <Dropdown
+                singleSelectOption
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/components/dropdown.md]: I think we should use `singleSelectOption={true}` since it's less confusing.  Some people who are new to React may confuse `singleSelectOption` with `singleSelectOption={singleSelectOption}`.

File: packages/wonder-blocks-dropdown/components/dropdown.test.js [hunk 5]
```diff
@@ -257,5 +257,63 @@ describe("Dropdown", () => {
...
+            const controlledComponent = mount(
+                <ControlledComponent singleSelectOption />,
+            );
+            controlledComponent.simulate("click");
+            controlledComponent.setState({selectedValues: ["B"]});
+            const dropdownCore = controlledComponent.find(DropdownCore);
+
+            // Act
+            dropdownCore.simulate("keydown", {keyCode: keyCodes.down});
+            dropdownCore.simulate("keyup", {keyCode: keyCodes.down});
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/components/dropdown.test.js]: This should probably be part of `// Arrange` since the thing that's being tested here is what happens when an option is selected.

File: packages/wonder-blocks-dropdown/components/dropdown.test.js [hunk 6]
```diff
@@ -257,5 +257,63 @@ describe("Dropdown", () => {
...
+
+            const optionItem = controlledComponent.find(OptionItem).at(0);
+            optionItem.simulate("click");
+
+            // Assert
+            expect(controlledComponent.state("selectedValues")).toHaveLength(1);
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/components/dropdown.test.js]: This indicates a slight inconsistency in the API between `SingleSelect` and `Dropdown` in select-select mode.  `SingleSelect`'s `onChange` passes back a single value where as `Dropdown`'s `onChange` passes an array even in single select mode.  I don't think this is something we need to solve in this PR.  Maybe this inconsistency is fine, but it is something to think about.

File: packages/wonder-blocks-dropdown/components/dropdown.js [hunk 7]
```diff
@@ -101,6 +107,12 @@ export default class Dropdown extends React.Component<Props, State> {
         };
     }
 
+    componentDidMount() {
+        if (this.props.opened != null) {
+            this.handleOpenChanged(this.props.opened);
+        }
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/components/dropdown.js]: This change feels pretty orthogonal to the single-select change and could've gone in a separate PR.

File: packages/wonder-blocks-clickable/docs.md (hunk 3 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-clickable/docs.md]: I wonder if it could be configured to only spell-check non-code parts of files and ignore words in the comments that are also in the code.

File: packages/wonder-blocks-dropdown/components/dropdown.md (hunk 4 above)
Comment (kumailn) [reply packages/wonder-blocks-dropdown/components/dropdown.md]: Sure, we also do something similar with `skipClientNav` which is why I did it here

File: packages/wonder-blocks-dropdown/components/dropdown.js (hunk 1 above)
Comment (kumailn) [reply packages/wonder-blocks-dropdown/components/dropdown.js]: @jandrade I agree, I’ve updated it in my latest commit to reflect your suggesiton
